twoliter/src/lock.rs

```diff
@@ -28,3 +42,239 @@ pub(crate) struct LockedImage {
     /// The resolved image uri of the dependency
     pub source: String,
 }
+
+impl LockedImage {
+    pub fn new(vendor: &Vendor, image: &Image) -> Self {
+        Self {
+            name: image.name.clone(),
+            version: image.version.clone(),
+            vendor: image.vendor.clone(),
+            source: format!("{}/{}:v{}", vendor.registry, image.name, image.version),
+        }
+    }
+}
+
+impl Display for LockedImage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!(
+            "{}-{}@{} ({})",
+            self.name, self.version, self.vendor, self.source,
+        ))
+    }
+}
+
+/// The hash should not contain the source to allow for collision detection
+impl Hash for LockedImage {
```

Comment:
I'm surprised clippy isn't bugging you about Eq and Ord being different from Hash. Seems fine? Ish?